### ISSUE TITLE
Implement PTLC scripts

### ISSUE THREAD

**Issue by GeneFerneau** [bitcoin-teleport/teleport-transactions#37]
Point Time Locked Contracts (PTLC) scripts allow for swaps to look like normal transactions. Especially with Taproot, Schnorr signatures and tapscripts greatly increase the ability to blend into the crowd (enhancing user privacy). PTLCs can also be implemented with ECDSA signatures, though the setup and cryptography are more complex.

PTLCs, by using adaptor signatures, also remove the necessity of manually transferring a secret key (`hash_preimage`) between the maker and taker. This simplifies, and increases the security of the protocol.

Here is a basic, work-in-progress outline of how a PTLC protocol would work with coinswaps.

# PTLC protocol

- (maker) create adaptor private key:
  - `adaptor <-$- Zq`
- (maker) combine adaptor public key with taker public key:
  - `pointlock_key = takerPub + adaptor*G`
- (maker) create redeemscript contract using `pointlock_key` and `timelock_key`
  - send `redeemscript`, `pointlock_key`, and `timelock_key` to taker
---
- (taker) create musig2 nonce
  - `taker_musig_nonce <-$- Zq`
- (taker) aggregate the nonces in the pre-session to create a musig2 session
- (taker) create partial musig2 signature over contract using `my_private_key`
  - `taker_pre_sig = session.partial_sign(&secp, &mut taker_musig_nonce, &keypair, &pre_session)`
  - send `taker_pre_sig` and `taker_musig_pubnonce` to maker
---
- (maker) verify taker's partial signature over contract
- (maker) aggregate the nonces in the pre-session to create a musig2 session
- (maker) create adaptor signature over the taker's pre-signature:
  - `adaptor_sig = musig::adapt(&secp, &taker_pre_sig, &adaptor, nonce_parity)`
  - send to taker
---
- (taker) extract adaptor public key:
  - `adaptor_pubkey = pointlock_key - takerPub`
- (taker) extract adaptor private key:
  - `adaptor = adaptor_sig - taker_pre_sig`
- (taker) verify that `adaptor_pubkey = adaptor*G`
- (taker) add `adaptor` to `my_private_key`, and sign the contract to retrieve coin:
  - `pointlock_privkey = adaptor + my_private_key`
  - `secp.sign(&contract_hash, &pointlock_privkey)`
  - broadcast tx

## Open questions

- should the signature using the `pointlock_key` include a contribution from the maker, other than the adaptor key?
- should only PTLCs be supported (replace HTLCs), or should both be supported?

**Comment by GeneFerneau** [comment]
I'm currently working on an implementation. This issue is to track progress, and open discussion over the implementation.

**Comment by chris-belcher** [comment]
If I understand right, your scheme only works if the coinswap addresses are schnorr/taproot addresses.

That would not provide us with a very big anonymity set. I expect adoption of taproot addresses to be very slow (more details here https://gist.github.com/chris-belcher/9144bd57a91c194e332fb5ca371d0964#ecdsa-2p third paragraph).

I think a much better way to use PTLCs is to use schnorr to replace what are currently called `contract_redeemscripts` in the code. That would mean the coinswap addresses could be ecdsa-2p, but the contract addresses would be schnorr, and so schnorr would only hit the blockchain in the backout situation and not in the normal siguation where the coinswap proceeds successfully.

Another thing, do you have any thoughts on the design for how the timelock branch would work in your design? i.e. what happens if one of the parties disappears leaving the coins stuck in the musig address.

**Comment by GeneFerneau** [comment]
> If routed coinswap has a small anonymity set, then essentially all privacy is lost.

Agreed, but using Schnorr-Taproot would be indistinguishable from all other Schnorr-Taproot transactions in the happy path (keypath spend).

> For more details for why this is true see https://zmnscpxj.github.io/bitcoin/multiswap.html

This is a critique of a particular CoinSwap implementation that does little to obfuscate payment amounts. With the tiered, routed CoinSwap protocol you've proposed in your developer notes, this would be less of a problem, no?

After a first reading, I don't see anything in ZmnSCPxj's critique that screams "never use Taproot, anonymity too smol".

> Because coinswap depends so much on having a high anonymity set for coinswap addresses, then I reckon your plan of using schnorr addresses as coinswap addresses is useless.

With plans to implement PTLCs for Lightning as well (ostensibly using Taproot), the anonymity set is potentially pretty large. Even with the TapScript timelock spend (revealing the script), we could converge on whatever convention Lightning uses, or some other standard yet to be proposed. Hopefully we can have standard PTLC to increase the anonymity set even more.

There are also very likely to be even more use cases for Taproot transactions, which would make distinguishing CoinSwap addresses even harder.

Hardly seems useless to me. Especially with keypath spends, the majority of the time a script spend path would never even be revealed. Something that is impossible with non-Taproot transactions.

> You design which uses OP_CSV in a tapscript has a downside that it will be visible to everyone when the timelock branch is used

Your current design with the `contract_redeemscript` uses `OP_CSV`, which is where I took the TapScript impl from, so I don't see your point.

Your script is visible to all from the moment the transaction hits the blockchain, not just when the timelock is used like in a TapScript spend.

> When I talked about ECDSA-2P/Schnorr-Taproot design I just meant a scheme where the coinswap addresses are ecdsa addresses because they use ecdsa-2p, but the contract transactions use schnorr instead of contract_redeemscript, and so use adaptor signatures instead of hashlocks and use nSequence instead of OP_CSV timelocks.

I have no idea how you would use ECDSA for the address, but then use Schnorr-Taproot for contract transactions. What does that even mean?

AFAIU, there is no way to mix the two in a single transaction. Are you talking about two separate transactions: one paying to an ECDSA-2p address, and one paying to a Schnorr address?

---

I'm not anti-ECDSA-2p, and have done work on implementing some of the cryptographic primitives used for the protocol. I think having ECDSA-2p for the short-term definitely has a much larger anonymity set, and is a good thing to use.

We should definitely implement ECDSA-2p PTLC contracts.

Providing a good use-case for Schnorr-Taproot PTLCs in CoinSwap will also help to increase Taproot usage, growing the anonymity set. I know it is somewhat of a chicken-and-egg problem, which is why I want to implement both, and give users the option to choose.

For users with heightened threat models, the reduced anonymity set of Schnorr probably isn't worth it, and they might prefer ECDSA-2p based PTLCs. For those with a lower threat model, the lower anonymity set of Schnorr based PTLCs may be acceptable. 

Informing users of the tradeoffs gives them the opportunity to make that choice.

**Comment by moneyball** [comment]
Thanks for the good thoughts @chris-belcher! I agree the non-LN motivation to adopt taproot will be weaker than segwit, but I think LN will see strong enough growth that overall the adoption of taproot will exhibit similar timeframes as segwit. I'd expect a 2-4 year timeframe for 30-40% of recent transactions being taproot. If you and the contributors to this project agree with that timeframe, then it is a question of what you believe the timeframe for CoinSwap launch and adoption would be. However, if you/other contributors think my estimates are too rosy, or at least simply just too uncertain, then it makes a stronger case for supporting ECDSA.

**Comment by GeneFerneau** [comment]
Updated the original post with a gist of the protocol.

Still working on details. Added input from related discussions.

Will continue to update as discussion progresses.
